packages/mobile/src/hooks/useTikTokAuth.ts

```diff
@@ -1,37 +1,110 @@
+  init as tikTokInit,
+  auth as tikTokAuth,
+  events as tikTokEvents
+} from 'react-native-tiktok'
 
 import { track, make } from 'app/services/analytics'
+import { dispatch } from 'app/store'
 import * as oauthActions from 'app/store/oauth/actions'
 import { Provider } from 'app/store/oauth/reducer'
 import { EventNames } from 'app/types/analytics'
 
-export const useTikTokAuth = (args: UseTikTokAuthArguments) => {
-  const dispatch = useDispatch()
+const authenticationUrl = `${Config.IDENTITY_SERVICE}/tiktok`
 
-  return createUseTikTokAuthHook({
-    authenticate: async () => {
-      return new Promise<Credentials>((resolve, reject) => {
-        track(
-          make({
-            eventName: EventNames.TIKTOK_START_OAUTH
-          })
-        )
-        const authenticationUrl = `${Config.IDENTITY_SERVICE}/tiktok`
-
-        dispatch(
-          oauthActions.requestNativeOpenPopup(
-            resolve,
-            reject,
-            authenticationUrl,
-            Provider.TIKTOK
-          )
+const canOpenTikTok = () => {
+  return Linking.canOpenURL('tiktok://app')
+}
+
+const authenticate = async (): Promise<Credentials> => {
+  track(
+    make({
+      eventName: EventNames.TIKTOK_START_OAUTH
+    })
+  )
+
+  // Perform WebView auth if TikTok is not installed
+  // TikTok LoginKit is supposed to handle this but it doesn't seem to work
+  if (!(await canOpenTikTok())) {
+    return new Promise((resolve, reject) => {
+      dispatch(
+        oauthActions.requestNativeOpenPopup(
+          resolve,
+          reject,
+          authenticationUrl,
+          Provider.TIKTOK
         )
+      )
+    })
+  }
+
+  // Need to set a csrf cookie because it is required for web
+  await CookieManager.set(Config.IDENTITY_SERVICE, {
+    name: 'csrfState',
+    value: 'true',
+    secure: true
+  })
+
+  tikTokInit(Config.TIKTOK_APP_ID)
+
+  return new Promise((resolve, reject) => {
+    const handleTikTokAuth = async (
+      code: string,
+      error: boolean | null,
+      errorMessage: string
+    ) => {
+      if (error) {
+        reject(new Error(errorMessage))
+      }
+
+      const response = await fetch(
+        `${Config.IDENTITY_SERVICE}/tiktok/access_token`,
+        {
+          credentials: 'include',
+          method: 'POST',
+          body: JSON.stringify({
+            code,
+            state: 'true'
+          }),
+          headers: {
+            'Content-Type': 'application/json'
+          }
+        }
+      )
+
+      if (!response.ok) {
+        reject(new Error(response.status + ' ' + (await response.text())))
+      }
+
+      const {
+        data: { access_token, open_id, expires_in }
+      } = await response.json()
+
+      resolve({
+        accessToken: access_token,
+        openId: open_id,
+        expiresIn: expires_in
       })
-    },
+    }
+
+    // Needed for Android
+    tikTokEvents.addListener('onAuthCompleted', (resp) => {
+      handleTikTokAuth(resp.code, !!resp.status, resp.status)
```

Comment:
Just to confirm -- `!!resp.status` tells us if there is an error and `resp.status` is the error message? So it's empty if the thing is successful?

Reply:
Yes, the status is `0` if successful and otherwise is a code representing the error https://developers.tiktok.com/doc/getting-started-android-handling-errors

This is only for android